### ISSUE TITLE
Genuine global (persistent) variables

### ISSUE THREAD

**Issue by systemdlete** [rsyslog/rsyslog#4790]
### Expected behavior
Variables that are persistent across various types of messages within a single ruleset (or maybe wider than that even)

### Actual behavior
So-called rsyslog "global" variables are really more like defaults.  These are also referred to, variously, as state variables, though they cannot be modified, making them nearly useless as state variables (unless there is missing docs that might explain how to use them appropriately as genuine state flags/variables).   They can only be set once, and there are strict rules pertaining to them.    (I am referring to $/ variables here).

### Steps to reproduce the behavior
Self-explanatory.

### Environment
- rsyslog version:  8.1901
- platform: devuan beowulf (possibly devuan chimaera and later also)

### Additional information
This is a request for a new feature in rsyslog, not a modification to any existing constructs or behaviors.

I would like to be able to express a global flag--at least within a single ruleset--that maintains state information.  For instance, I might like to set a flag upon encountering a specific message pattern for later processing; then, later, when encountering a related message pattern pertaining to the same message-derived objects, check the flag state to address appropriate treatment for that message.

The instance I am considering, in particular, though it might equally apply to other conditions and situations, is where rsyslog sends a check to a nagios-style system for an interface that has reset for whatever reason.   At that point, the handling code for that message would simply set a state flag for the host/interface involved.   When rsyslog then sends a check for the interface having recovered from the temporary down state (usually only seconds), it would check that state flag so it would know to craft the check output to indicate that the interface is well again, but it had had a momentary outage.   This is just to make the operator aware that even though, overall, the interface is operating normally, there had been some issue that caused a reset and perhaps calls for investigation or mitigation.

This is not the most important issue in the systems monitoring universe, but having some kind of genuine state variables would be very helpful.    In addition, it would also be helpful to thoroughly review and update the docs so that current "global" or "state" variables are called something closer to their actual functionality, that of one-shot defaults, and save the terminology "state variable" for actual such.

**Comment by davidelang** [comment]
These already exist, $! and !. variables have the life of the message, $\ 
variables persist across messages and rulesets (but have a rather significant 
performance hit to use)

David Lang

**Comment by systemdlete** [comment]
davidelang:  The documentation (if you can call it that, a seemingly forgotten blog post from 2013) on $/  (not $\) variables indicates these can only be set once.  Actual experiments are true to the description in that blog post.

Also, keep in mind I am using a 2-year old version of rsyslog.   Perhaps behavior/facilities have changed since then, idk.

**Comment by davidelang** [comment]
that was not the intent on them, we'll see what Rainer has to say on it, but I 
think it's more likely that $/ will get fixed (sorry for remembering the wrong 
direction) rather than something new added.

since the three classes of variables are all single json entities under the 
covers, it doesn't make sense that they would only be able to get set once.

you will need to get onto a current version to take advantage of any new feature 
added (and without the ability to test a version you have asked for, odds of 
getting effort on it are lower)

David Lang

**Comment by systemdlete** [comment]
I'll be upgrading to a newer devuan soon.

Here is the description from the blog post:   https://rainer.gerhards.net/2013/10/a-proposal-for-rsyslog-state-variables.html.    David, if this is also in current documentation, please provide a direct link so I can review it for myself.  It does not seem to appear in the 8.1901 documentation, so far as I have been able to search through it.  Maybe I missed it.

**Comment by systemdlete** [comment]
Here is a short "proof" script I wrote to demonstrate this (upload would not work for my .conf file):

`module(load="imudp")
input(type="imudp" port="513" ruleset="test")

template(name="debug-trace" type="list") {
	constant(value="DEBUG: '")
	property(name="$.z!debug")
	constant(value="'\n")
}

ruleset(name="test")
{
	unset $.z;

	set $.z!debug = ">>> " & $msg;
	action(type="ompipe" pipe="/tmp/mytest" template="debug-trace")

	# "e1000 0000:00:08.0 eth1: Reset adapter"
	if(re_match($msg,
		"[A-Za-z0-9]+ [0-9a-f:.]+ (eth[0-9]+): Reset adapter"
	)) then {
		set $.x!hostname = $hostname;
		set $.x!interface = re_extract($msg,
			"[A-Za-z0-9]+ [0-9a-f:.]+ (eth[0-9]+): Reset adapter"
			,0,1,"");
		set $.z!debug = "host: " & $.x!hostname & ", interface: " & $.x!interface;
		set $/state = $.x;

		set $.z!debug = "showing key-values of reset state";
		action(type="ompipe" pipe="/tmp/mytest" template="debug-trace")
		foreach ($.i in $/state) do
		{
			set $.z!debug = $.i!key & ":" & $.i!value;
			action(type="ompipe" pipe="/tmp/mytest" template="debug-trace")
		}
	}
}`

Run this as "RSYSLOG_DEBUG=DebugOnDemand rsyslogd -n -f test.conf.globals -i /tmp/junk.pid"
(I am assuming the above rsyslog config file is in test.conf.globals)

On the rsyslog server system, in a different window perhaps, run:
logger -n $(hostname) -P 513 "e1000 0000:00:08.0 eth1: Reset adapter"

Then on a different system, send the rsyslog server:
logger -n $(hostname) -P 513 "e1000 0000:00:08.0 eth1: Reset adapter"

Hopefully, this will either demonstrate the problem, or else prove that it has changed in a newer version you are using.

**Comment by systemdlete** [comment]
BTW, thank you for your very prompt replies.  If you want any additional supporting info, let me know.

(Also, if you have tips for formatting my last post, please advise!  Thanks.)

**Comment by davidelang** [comment]
simple test

module(load="imuxsock") # provides support for local system logging
module(load="imklog" permitnonkernelfacility="on")
$ActionFileDefaultTemplate RSYSLOG_TraditionalFileFormat
$WorkDirectory /var/spool/rsyslog
/var/log/messages
if $programname == "set" then {
     set $/test = $msg;
     /var/log/set;RSYSLOG_DebugFormat

logger -t set test1
logger -t set test2

cat /var/log/set

Debug line with all properties:
FROMHOST: 'dlang-mobile', fromhost-ip: '127.0.0.1', HOSTNAME: 'dlang-mobile', 
PRI: 13,
syslogtag 'set:', programname: 'set', APP-NAME: 'set', PROCID: '-', MSGID: '-',
TIMESTAMP: 'Jan 29 22:22:37', STRUCTURED-DATA: '-',
msg: ' test1'
escaped msg: ' test1'
inputname: imuxsock rawmsg: '<13>Jan 29 22:22:37 set: test1'
$!:
$.:
$/:{ "test": " test1" }

Debug line with all properties:
FROMHOST: 'dlang-mobile', fromhost-ip: '127.0.0.1', HOSTNAME: 'dlang-mobile', 
PRI: 13,
syslogtag 'set:', programname: 'set', APP-NAME: 'set', PROCID: '-', MSGID: '-',
TIMESTAMP: 'Jan 29 22:22:50', STRUCTURED-DATA: '-',
msg: ' test2'
escaped msg: ' test2'
inputname: imuxsock rawmsg: '<13>Jan 29 22:22:50 set: test2'
$!:
$.:
$/:{ "test": " test2" }

This shows that the variable is getting changed

now, there are race conditions (batches, multiple threads, etc) that can cause 
the change to not be visible to other messages processed very close to the time 
that the change happens.

David Lang

**Comment by systemdlete** [comment]
The question here is not whether the value of one key:value pair gets updated, but whether additional key:value pairs appear in the $/ variable.   Maybe you can change your script a little to model that scenario, which is closer to mine?

**Comment by davidelang** [comment]
module(load="imuxsock") # provides support for local system logging
module(load="imklog" permitnonkernelfacility="on")
$ActionFileDefaultTemplate RSYSLOG_TraditionalFileFormat
$WorkDirectory /var/spool/rsyslog
/var/log/messages
if $programname == "set" then {
     set $/test = $msg;
     /var/log/set;RSYSLOG_DebugFormat
if $programname == "set2" then {
     set $/test2 = $msg;
     /var/log/set;RSYSLOG_DebugFormat



   788  rsyslogd -f /etc/rsyslog.conf
   789  logger te4st
   790  logger -t set test1
   791  logger -t set test2
   792  logger -t set2 test23
   793  cat /var/log/set



Debug line with all properties:
FROMHOST: 'dlang-mobile', fromhost-ip: '127.0.0.1', HOSTNAME: 'dlang-mobile', 
PRI: 13,
syslogtag 'set:', programname: 'set', APP-NAME: 'set', PROCID: '-', MSGID: '-',
TIMESTAMP: 'Jan 30 02:32:01', STRUCTURED-DATA: '-',
msg: ' test1'
escaped msg: ' test1'
inputname: imuxsock rawmsg: '<13>Jan 30 02:32:01 set: test1'
$!:
$.:
$/:{ "test": " test1" }

Debug line with all properties:
FROMHOST: 'dlang-mobile', fromhost-ip: '127.0.0.1', HOSTNAME: 'dlang-mobile', 
PRI: 13,
syslogtag 'set:', programname: 'set', APP-NAME: 'set', PROCID: '-', MSGID: '-',
TIMESTAMP: 'Jan 30 02:32:09', STRUCTURED-DATA: '-',
msg: ' test2'
escaped msg: ' test2'
inputname: imuxsock rawmsg: '<13>Jan 30 02:32:09 set: test2'
$!:
$.:
$/:{ "test": " test2" }

Debug line with all properties:
FROMHOST: 'dlang-mobile', fromhost-ip: '127.0.0.1', HOSTNAME: 'dlang-mobile', 
PRI: 13,
syslogtag 'set2:', programname: 'set2', APP-NAME: 'set2', PROCID: '-', MSGID: 
'-',
TIMESTAMP: 'Jan 30 02:32:16', STRUCTURED-DATA: '-',
msg: ' test23'
escaped msg: ' test23'
inputname: imuxsock rawmsg: '<13>Jan 30 02:32:16 set2: test23'
$!:
$.:
$/:{ "test": " test2", "test2": " test23" }

**Comment by davidelang** [comment]
I would suggest setting up a RSYSLOG_DebugFormat log file and look at the case 
where a message doesn't have the result that you are expecting. 90+% of the 
time, the problem is that some content doesn't match what you are expecting

David Lang

**Comment by rgerhards** [comment]
Global variables are global.  However, as @davidelang said, most things inside rsyslog run in parallel. There is no guarantee of order of exectution. If you do things other than counting, the results may be surprising - but still right.

In you sample, you use message-local variables. Global vars start with `$/`. Maybe that's the problem.

**Comment by davidelang** [comment]
testing assigning a hierarchy

this does surprise me in that it assigned it as a string rather than json, but 
that could be how $jsonmesg works (I normally use $! in templates, but that's 
not accepted, trying $all-json next

David Lang

module(load="imuxsock") # provides support for local system logging
module(load="imklog" permitnonkernelfacility="on")
$ActionFileDefaultTemplate RSYSLOG_TraditionalFileFormat
$WorkDirectory /var/spool/rsyslog
/var/log/messages
if $programname == "set" then {
     set $/test = $msg;
     /var/log/set;RSYSLOG_DebugFormat
if $programname == "set2" then {
     set $/test2 = $jsonmesg;
     /var/log/set;RSYSLOG_DebugFormat


   818  logger test
   819  tail /var/log/messages
   820  logger -t set test2
   821  logger -t set2 test3
   822  cat /var/log/set



Debug line with all properties:
FROMHOST: 'dlang-mobile', fromhost-ip: '127.0.0.1', HOSTNAME: 'dlang-mobile', 
PRI: 13,
syslogtag 'set:', programname: 'set', APP-NAME: 'set', PROCID: '-', MSGID: '-',
TIMESTAMP: 'Jan 30 02:46:09', STRUCTURED-DATA: '-',
msg: ' test2'
escaped msg: ' test2'
inputname: imuxsock rawmsg: '<13>Jan 30 02:46:09 set: test2'
$!:
$.:
$/:{ "test": " test2" }

Debug line with all properties:
FROMHOST: 'dlang-mobile', fromhost-ip: '127.0.0.1', HOSTNAME: 'dlang-mobile', 
PRI: 13,
syslogtag 'set2:', programname: 'set2', APP-NAME: 'set2', PROCID: '-', MSGID: 
'-',
TIMESTAMP: 'Jan 30 02:46:17', STRUCTURED-DATA: '-',
msg: ' test3'
escaped msg: ' test3'
inputname: imuxsock rawmsg: '<13>Jan 30 02:46:17 set2: test3'
$!:
$.:
$/:{ "test": " test2", "test2": "{ \"msg\": \" test3\", \"rawmsg\": \"<13>Jan 30 
02:46:17 set2: test3\", \"timereported\": \"2022-01-30T02:46:17.157662-08:00\", 
\"hostname\": \"dlang-mobile\", \"syslogtag\": \"set2:\", \"inputname\": 
\"imuxsock\", \"fromhost\": \"dlang-mobile\", \"fromhost-ip\": \"127.0.0.1\", 
\"pri\": \"13\", \"syslogfacility\": \"1\", \"syslogseverity\": \"5\", 
\"timegenerated\": \"2022-01-30T02:46:17.157662-08:00\", \"programname\": 
\"set2\", \"protocol-version\": \"0\", \"structured-data\": \"-\", \"app-name\": 
\"set2\", \"procid\": \"-\", \"msgid\": \"-\", \"uuid\": null, \"$!\": null }" }

**Comment by davidelang** [comment]
$all-json also not accepted, since it's almost 3am here I'm going to give up for 
the night.

David Lang
kk

**Comment by rgerhards** [comment]
> testing assigning a hierarchy

I need to check deep. Out of my head, assigning hierarchies was never intended. The JSON part has become quite a bit abused in my PoV.

**Comment by davidelang** [comment]
On Sun, 30 Jan 2022, Rainer Gerhards wrote:

>> testing assigning a hierarchy
>
> I need to check deep. Out of my head, assigning hierarchies was never intended. The JSON part has become quite a bit abused in my PoV.

I've done that a lot in the $! and $. namespaces (I setup a $!trusted!path that 
contains info on the most recent hop (timestamp, fromhost-ip, input, hostname of 
receiving system), and then when it arrives on the next hop, that gets assigned 
to $!trusted!path!lasthop (through another variable) so that I have all the 
details of the path through multiple servers.

David Lang

**Comment by systemdlete** [comment]
Please look at the example I gave.  There is indeed a state variable I am using called '$/state'.

The examples you have responded with do not indicate which version of rsyslog you are using.   Are you stating that there HAD been a bug that has since been corrected?   I understand I am using a slightly old version, but that is the one that comes with Devuan Beowulf (~debian 10).   I have not yet tried this on Devuan Chimaera (~debian 11).

But at any rate, how can you realistically call these state variables when, due to implementation issues (parallel tasks), it would be impossible to use them as genuine state variables.   The state of any kind of object could change many, many times over the life of a process, and certainly would in most instances.  It sounds like it would be a good idea to call these counters or something else.   The term "global" means something entirely else, at least on this side of the pond.  (You would have to implement some kind of locking AND sequencing to make this work.)

This is not a criticism of rsyslog design, but tbh, it is sounding as if there will never be any true state variables as understood generally for programming implemented for rsyslog.   If so, I may need to look at some other means of accomplishing what I am looking to do.   Is there a way to do what I need by somehow "pushing" a message on a queue (or maybe another construct) to be processed later?  It seems like this could almost be possible in rsyslog, but I only saw bits of examples that might get me there, but no complete example of this sort of approach.

**Comment by davidelang** [comment]
they are truly global, but the order of processing isn't first-in-first-out, so 
rsyslog may be processing a message that arrived later before the variable is 
set by a message that arrived earlier. The path that a message follows, 
especially when relayed between machines, can be far more convoluted than is 
obvious at first glance. I don't know that that's what's happening in your case 
(I don't understand your case yet). but the examples that I provided do show 
that the value of the $/ variables are getting changed and multiple variables 
are getting set (and reset)

My examples are on the default rsyslog that shiped with ubuntu 21.10 (8.2102)

Your example is rather complex, and it's not clear what the messages that you 
are receiving look like and what you expect to take place. Can you try to 
simplify your test case and provide the log messages that you expect to trigger 
changes (and show what you expect to see where it differs from what you do see). 
As I said before DebugFormat output is very useful in tracking down unexpected 
behavior.

David Lang

**Comment by rgerhards** [comment]
> Please look at the example I gave. There is indeed a state variable I am using called '$/state'.

Indeed I had to search to see it ;-)

But now I am even more puzzled. Why do you use a global variable at all? It looks like it is just used as a temporary/local variable in the context of a specific message and only when some condition is met. Can you explain?

**Comment by systemdlete** [comment]
The same message can capture more than one different interface.  E.g., it might be eth0, eth1, eth2, etc.  Each needs to be tracked individually.   A reset on one adapter is not necessarily related to a reset on another.  This is why the use of regular expression there.   The format of an example message is immediately above the if statement.

I provided this example as an excerpt from a larger script, trying to give you a short proof.   I did not write a general case from scratch.  I even gave you all instructions on how to recreate the scenario so you could see for yourself.  The proof I offered uses the newer syntax, which per your own documentation, is preferable most of the time.  I thought I was doing everything properly.

As far as re-writing another version, I am not sure if I could re-create the problem that way.   Why not try the example that I did give you, even if you use your most current version of rsyslog?   That way you can see what I am seeing (unless your newer version has eliminated this bug, which would be nice).

Would you kindly respond to my other issues with rsyslog (1) concerning using some sort of diversion queue, or some other solution?  And (2) my request to update your documentation re the use of the term "state variable" -- yes, your global may indeed be truly global, but your global is not a true state variable because it cannot be relied upon as such, for valid technical/implementation reasons, by your own statements here in this thread.

**Comment by systemdlete** [comment]
I should probably add that when I mentioned "documentation," I meant--as I pointed out previously--there is a dearth of information about these global variables.  The only info I have ever encountered was that link to a very old blog entry on a site that seems to be defunct.   I ONLY encountered it because I happened upon another post on  different site that linked to it.   Otherwise, I did not see doc for globals in the 8.1901 documentation (but if I missed it, please provide a link to it, thanks).

**Comment by davidelang** [comment]
I thought all three types were defined together, I see that global variables 
have not made it into the documentation at all, so Rainer will have to comment 
on how they actually work (or we'll have to dig into the source)

> I should probably add that when I mentioned "documentation," I meant--as I pointed out previously--there is a dearth of information about these global variables.  The only info I have ever encountered was that link to a very old blog entry on a site that seems to be defunct.   I ONLY encountered it because I happened upon another post on  different site that linked to it.   Otherwise, I did not see doc for globals in the 8.1901 documentation (but if I missed it, please provide a link to it, thanks).

David Lang

**Comment by davidelang** [comment]
I'm still struggling to understand what you are trying to do.

Can you provide a sample log line and an example of what you think the resulting 
json variable contents should look like?

are you trying to add elements to a json array? If so, that could be the 
problem. Rsyslog has very poor support for arrays (pretty much only creating 
them via a mm module like  mmjsonparse or mmnormalize and walking them with 
foreach)

It could be that your real need is more around better json array support than 
anything else

David Lang

**Comment by systemdlete** [comment]
David:  Please read my original posts.   I provided an example of the type of log line, and I even gave examples of logger commands with arguments to run to re-create the scenario.    I have even re-iterated this in subsequent posts here.

Maybe start at the beginning of the thread and work your way forward.   If there is something that is not clear, please point to the specific post in this thread that confuses you.  But before doing so, please double-check if I might have specified what you want to know in a preceding post.

**Comment by davidelang** [comment]
I assume this is the 'fist message' you are telling us to read? (even though 
it's not the first)

On Sat, 29 Jan 2022, systemdlete wrote:

> Here is a short "proof" script I wrote to demonstrate this (upload would not work for my .conf file):
>
> `module(load="imudp")
> input(type="imudp" port="513" ruleset="test")
>
> template(name="debug-trace" type="list") {
> 	constant(value="DEBUG: '")
> 	property(name="$.z!debug")
> 	constant(value="'\n")
> }
>
> ruleset(name="test")
> {
> 	unset $.z;
>
> 	set $.z!debug = ">>> " & $msg;
> 	action(type="ompipe" pipe="/tmp/mytest" template="debug-trace")
>
> 	# "e1000 0000:00:08.0 eth1: Reset adapter"
> 	if(re_match($msg,
> 		"[A-Za-z0-9]+ [0-9a-f:.]+ (eth[0-9]+): Reset adapter"
> 	)) then {
> 		set $.x!hostname = $hostname;
> 		set $.x!interface = re_extract($msg,
> 			"[A-Za-z0-9]+ [0-9a-f:.]+ (eth[0-9]+): Reset adapter"
> 			,0,1,"");

so at this point $.x is {'hostname' : 'some name', 'interface': 'ethN'}

> 		set $.z!debug = "host: " & $.x!hostname & ", interface: " & $.x!interface;
> 		set $/state = $.x;

at this point $/state is either '' or equal to $.x (depending on if you can 
assign json structures to state variables)

> 		set $.z!debug = "showing key-values of reset state";
> 		action(type="ompipe" pipe="/tmp/mytest" template="debug-trace")
> 		foreach ($.i in $/state) do
>
> 		{
> 			set $.z!debug = $.i!key & ":" & $.i!value;

$.z is either '' (if the assignment of the structure fails), 
'hostname:somename', or 'interface:ethn'

when the next log message is processed, nothing looks at the old value of 
$/state, so this example makes no use of the global feature of $/ variables and 
only suffers the drawbacks of using $/state instead of $.state.

I have shown you examples of $/ variables getting different values at different 
times.

The only think I haven't done is try to assign a json structure to a global 
variable.

David Lang

> 			action(type="ompipe" pipe="/tmp/mytest" template="debug-trace")
> 		}
> 	}
> }`
>
> Run this as "RSYSLOG_DEBUG=DebugOnDemand rsyslogd -n -f test.conf.globals -i /tmp/junk.pid"
> (I am assuming the above rsyslog config file is in test.conf.globals)
>
> On the rsyslog server system, in a different window perhaps, run:
> logger -n $(hostname) -P 513 "e1000 0000:00:08.0 eth1: Reset adapter"
>
> Then on a different system, send the rsyslog server:
> logger -n $(hostname) -P 513 "e1000 0000:00:08.0 eth1: Reset adapter"
>
> Hopefully, this will either demonstrate the problem, or else prove that it has changed in a newer version you are using.
>
>

**Comment by davidelang** [comment]
module(load="imuxsock") # provides support for local system logging
module(load="imklog" permitnonkernelfacility="on")
$ActionFileDefaultTemplate RSYSLOG_TraditionalFileFormat
$WorkDirectory /var/spool/rsyslog
/var/log/messages
if $programname == "set" then {
     set $/test = $msg;
     set $.test3!host = $hostname;
     set $.test3!msg = $msg;
     set $/test3 = $.test3;
     set $/test4!host = $hostname;
     set $/test4!msg = $msg;
     /var/log/set;RSYSLOG_DebugFormat
if $programname == "set2" then {
     set $/test2 = $!all-json;
     /var/log/set;RSYSLOG_DebugFormat


   862  logger test
   863  logger -t set2 test2
   864  logger -t set3 test3
   867  logger -t set test3

Jan 31 18:54:28 dlang-mobile rsyslogd: [origin software="rsyslogd" 
swVersion="8.2102.0" x-pid="2257157" x-info="https://www.rsyslog.com"] start
Jan 31 18:54:32 dlang-mobile dlang: test
Jan 31 18:54:38 dlang-mobile set2: test2
Jan 31 18:54:44 dlang-mobile set3: test3
Jan 31 18:55:29 dlang-mobile set: test3

***@***.***:~# cat /var/log/set
Debug line with all properties:
FROMHOST: 'dlang-mobile', fromhost-ip: '127.0.0.1', HOSTNAME: 'dlang-mobile', 
PRI: 13,
syslogtag 'set2:', programname: 'set2', APP-NAME: 'set2', PROCID: '-', MSGID: 
'-',
TIMESTAMP: 'Jan 31 18:54:38', STRUCTURED-DATA: '-',
msg: ' test2'
escaped msg: ' test2'
inputname: imuxsock rawmsg: '<13>Jan 31 18:54:38 set2: test2'
$!:
$.:
$/:{ "test2": "" }

Debug line with all properties:
FROMHOST: 'dlang-mobile', fromhost-ip: '127.0.0.1', HOSTNAME: 'dlang-mobile', 
PRI: 13,
syslogtag 'set:', programname: 'set', APP-NAME: 'set', PROCID: '-', MSGID: '-',
TIMESTAMP: 'Jan 31 18:55:29', STRUCTURED-DATA: '-',
msg: ' test3'
escaped msg: ' test3'
inputname: imuxsock rawmsg: '<13>Jan 31 18:55:29 set: test3'
$!:
$.:{ "test3": { "host": "dlang-mobile", "msg": " test3" } }
$/:{ "test2": "", "test": " test3", "test3": { "host": "dlang-mobile", "msg": " 
test3" }, "test4": { "host": "dlang-mobile", "msg": " test3" } }

this says that you can assign json objects to global variables and can assign 
items to multi-layer global variables

and doing a logger -t set testtest results in

Debug line with all properties:
FROMHOST: 'dlang-mobile', fromhost-ip: '127.0.0.1', HOSTNAME: 'dlang-mobile', 
PRI: 13,
syslogtag 'set:', programname: 'set', APP-NAME: 'set', PROCID: '-', MSGID: '-',
TIMESTAMP: 'Jan 31 19:03:59', STRUCTURED-DATA: '-',
msg: ' testtest'
escaped msg: ' testtest'
inputname: imuxsock rawmsg: '<13>Jan 31 19:03:59 set: testtest'
$!:
$.:{ "test3": { "host": "dlang-mobile", "msg": " testtest" } }
$/:{ "test2": "", "test": " testtest", "test3": { "host": "dlang-mobile", "msg": 
" test3" }, "test4": { "host": "dlang-mobile", "msg": " testtest" }, "host": 
"dlang-mobile", "msg": " testtest" }

can you please add some RSYSLOGDebugFormat output messages to your test and see 
what you are getting at different stages?

David Lang

**Comment by systemdlete** [comment]
Try using the new syntax.   Try using my example?   Please?   Preferably v 8.1901, or indicate what version you are using.

Maybe the problem occurs with the new syntax.  I don't know.

**Comment by davidelang** [comment]
what part of the new syntax do you think I am not using?

**Comment by davidelang** [comment]
going back to this original message, you seem very concerned by what you find in 
the documentation of the proposed feature (before it was implemented), but you 
don't actually say what happens when you try it or what you expect the results 
to be 'self-explanatory' is not a good statement when running into other 
examples where it does work as expected)

note, I am just another user trying to help you (I do happen to be the person 
who proposed having three types of variables, although in practice I haven't had 
a reason to use the global variables yet, other than fairly toy implementation 
to turn sets of logging on and off for debugging purposes)

David Lang

**Comment by rgerhards** [comment]
@systemdlete sorry, but free software is not "free as in free beer". I don't owe you anything and don't find your request and wording decent. After all, it's my free time I am working on this project and I am not obliged to make it as easy as possible for you. 

The bottom line is that what you try to do does not work in a world where everything runs in parallel. You may configure rsyslog to do everything sequential (single ruleset, one thread). 

@davidelang indeed, something with the doc smell fishy. Not sure if we lost some update. I am pretty sure I had written up on the different types of variables.

**Comment by davidelang** [comment]
On Tue, 1 Feb 2022, Rainer Gerhards wrote:

> @davidelang indeed, something with the doc smell fishy. Not sure if we lost some update. I am pretty sure I had written up on the different types of variables.

I remember seeing some page where the three types were defined, but searching 
the docs I don't find the string $/ except in a couple examples of rsyslog 
output.

David Lang

**Comment by rgerhards** [comment]
@davidelang me, too - and that's strange. Not sure yet if it is faster to dig through the repo or re-write it...

**Comment by davidelang** [comment]
probably re-write it. We have a LOT more experience with things since the first 
time. Not to mention the results of refactoring and cleanups over the years.

how many of the limitations from the initial draft proposal are still in the 
final code? $/ is clearly a json hierarchy, unlike the draft, and the normal 
json manipulation tools work.

How many of the special functions actually got implemented?

May be a chance to clean up the entire section on properties/variables, how to 
reference them (both old and new syntax), etc.

David Lang

**Comment by rgerhards** [comment]
@davidelang we should probably add a separate issue for this, but a quick reply:

IIRC the special functions were not implemented because there was never real need expressed. The restrictions are still there,. because it simply is impossible to provide any other guarantee. Without workflow-dependent mutex locks. In other words: we would need to add mutex calls to the script, so that the user can configure them at the right places. Almost all folks will get it wrong (it's hard!) as you need a good parallel code development education to do it right. And this IMHO is definitely ways out of scope of a config language. If such a thing is desired, this should be done in a separate process (like SEC, for example).

**Comment by systemdlete** [comment]
I wish I had started with this approach when I first logged this issue.   I think it more concisely presents the problem I am trying to solve:   [https://stackoverflow.com/questions/68655622/how-to-correctly-handle-and-parse-multiline-logs-from-rsyslog](url)

I think this is a perfectly reasonable request.  This is a capability that admins might need.   (BTW, syslog-ng is able to do something like this, iirc).

**Comment by davidelang** [comment]
rsyslog does not try to handle logs where different log entries are split and 
mixed with each other.

If you are dealing with logs where the logs are not written out atomically, then 
what's to say that they are split at the end of lines? why would they not get 
split in the middle of a line?

trying to reassemble logs after they have gotten split and the bits intermingled 
with other logs seens liken an impossible task as a general problem.

Having multiple programs write to the same file is just begging for the log 
lines to get intermingled in undetectable ways (as a general thing if you do it 
on linux, the separate programs/threads will generally get their outputs 
combined on 4k boundries, not nicely on newline boundries)

David Lang
